src/Grpc.AspNetCore.Server/Internal/GrpcServiceBinder.cs

```diff
@@ -18,47 +18,98 @@
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ServerStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateServerStreaming(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, UnaryServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateUnary(method);
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
+        }
+
+        private void AddMethodCore(IMethod method, RequestDelegate requestDelegate)
+        {
+            ServiceMethods.Add(method);
+            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, requestDelegate, new object[] { method }));
+        }
+
+        internal void CreateUnimplementedEndpoints()
+        {
+            // Create endpoint that will return UNIMPLEMENTED status for missing method:
+            // - /Package.Service/{method}
+            // Create endpoint that will return UNIMPLEMENTED status for missing service:
+            // - /Package.{service}/{method}
+
+            // Typically there should be one service name for a type
+            // In case the bind method sets up multiple services in one call we'll loop over them
+            var serviceNames = ServiceMethods.Select(m => m.ServiceName).Distinct().ToList();
+
+            foreach (var serviceName in serviceNames)
+            {
+                if (_serviceMethodsRegistry.Methods.Any(m => string.Equals(m.ServiceName, serviceName, StringComparison.Ordinal)))
+                {
+                    // Unimplemented endpoint has already been added for this service
+                    continue;
+                }
+
+                EndpointConventionBuilders.Add(_builder.MapPost(serviceName + "/{unimplementedMethod}", _serverCallHandlerFactory.CreateUnimplementedMethod()));
+
+                var packageDelimiterIndex = serviceName.LastIndexOf('.');
+                if (packageDelimiterIndex != -1)
+                {
+                    var packageName = serviceName.Substring(0, packageDelimiterIndex + 1);
+
+                    // Unimplemented endpoint has already been added for this package
+                    if (_serviceMethodsRegistry.Methods.Any(m => m.ServiceName.StartsWith(packageName, StringComparison.Ordinal)))
+                    {
+                        continue;
+                    }
+
+                    EndpointConventionBuilders.Add(_builder.MapPost(packageName + "{unimplementedService}/{unimplementedMethod}", _serverCallHandlerFactory.CreateUnimplementedService()));
```

Comment:
Hmm now that I think of it, is the prefix sufficient to avoid conflicts with other non-gRPC endpoints? I'm not familiar with endpoints routing but what happens if multiple patterns are matched?

Reply:
There is route precedence.

"Greet.Greeter/SayHello" is matched by both of these routes:

Route 1: "Greet.Greeter/SayHello"
Route 2: "{service}/{method}"

But the first one has a higher precedence because it matches literals instead of parameters.

Reply:
> I'm not familiar with endpoints routing but what happens if multiple patterns are matched for example?

If two routes are matched and they have the same precedence then an error is thrown.

Reply:
Sorry, I was thinking of another scenario. See the updated comment.

Reply:
So a more concrete example:
Endpoint 1:
- Mapping a gRPC service called Namespace.GrpcService
- Creates a mapping of Namespace.*/* => unimplemented service request delegate
Endpoint 2:
- Mapping a static file server at Namespace.StaticServer
- Creates a mapping of Namespace.StaticServer/* => files

A request comes in with /Namespace.StaticServer/file.txt. This matches both routes and both would be considered parameters right? Then an error will be thrown?

Reply:
If someone had `/Namespace.StaticServer/file.txt` as a route then that would match because it is a literal. It would only be a problem if it had the same literal prefix followed by two parameters.

Note: static files don't (go through routing.

Reply:
Doesn't have to be static files, any other non-gRPC request would suffice. I suppose it's fine to add this now and if we ever run into mapping issues, we'll look into alternatives.